kroxylicious/src/main/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilter.java

```diff
@@ -0,0 +1,224 @@
+        response.groups().forEach(responseGroup -> responseGroup.topics().forEach(topic -> removeTenantPrefix(context, topic::name, topic::setName, false)));
+        context.forwardResponse(response);
+    }
+
+    @Override
+    public void onOffsetForLeaderEpochRequest(OffsetForLeaderEpochRequestData request, KrpcFilterContext context) {
+        request.topics().forEach(topic -> applyTenantPrefix(context, topic::topic, topic::setTopic, false));
+        context.forwardRequest(request);
+    }
+
+    @Override
+    public void onOffsetForLeaderEpochResponse(OffsetForLeaderEpochResponseData response, KrpcFilterContext context) {
+        response.topics().forEach(topic -> removeTenantPrefix(context, topic::topic, topic::setTopic, false));
+        context.forwardResponse(response);
+    }
+
+    @Override
+    public void onOffsetCommitRequest(OffsetCommitRequestData request, KrpcFilterContext context) {
+        request.topics().forEach(topic -> applyTenantPrefix(context, topic::name, topic::setName, false));
+        context.forwardRequest(request);
+    }
+
+    @Override
+    public void onOffsetCommitResponse(OffsetCommitResponseData response, KrpcFilterContext context) {
+        response.topics().forEach(topic -> removeTenantPrefix(context, topic::name, topic::setName, false));
+        context.forwardResponse(response);
+    }
+
+    @Override
+    public void onFetchRequest(FetchRequestData request, KrpcFilterContext context) {
+        request.topics().forEach(topic -> applyTenantPrefix(context, topic::topic, topic::setTopic, topic.topicId() != null));
+        context.forwardRequest(request);
+    }
+
+    @Override
+    public void onFetchResponse(FetchResponseData response, KrpcFilterContext context) {
+        response.responses().forEach(topic -> removeTenantPrefix(context, topic::topic, topic::setTopic, topic.topicId() != null));
+        context.forwardResponse(response);
+
+    }
+
+    private void applyTenantPrefix(KrpcFilterContext context, Supplier<String> getter, Consumer<String> setter, boolean ignoreEmpty) {
+        var tenantPrefix = getTenantPrefix(context);
+        String clientSideName = getter.get();
+        if (ignoreEmpty && (clientSideName == null || clientSideName.isEmpty())) {
+            return;
+        }
+        setter.accept(String.format("%s-%s", tenantPrefix, clientSideName));
+    }
+
+    private void removeTenantPrefix(KrpcFilterContext context, Supplier<String> getter, Consumer<String> setter, boolean ignoreEmpty) {
+        var tenantPrefix = getTenantPrefix(context);
+        var brokerSideName = getter.get();
+
+        if (ignoreEmpty && (brokerSideName == null || brokerSideName.isEmpty())) {
+            return;
+        }
+
+        setter.accept(brokerSideName.substring(tenantPrefix.length() + 1));
+    }
+
+    private static String getTenantPrefix(KrpcFilterContext context) {
+        return context.sniHostname().split("\\.")[0];
```

Comment:
Doing an `indexOf()` and `substring()` will avoid temp Pattern creation and allocations for all the DNS components required for `split()`.